Allow consuming std::array's; allow other output types in consume

jagerman opened the pull request:
This rewrites the tuple/pair code a bit so that it is also usable for std::array so that things like:

    auto arr = btdp.consume_list<std::array<int, 3>>("v");

now works.  Previously we could consume to a tuple, pair, or expandable container like a vector, but not to an std::array.  Now we can.

It also enlarges the recognized things in `consume` to allow any arbitrary deserializable list/dict types (including tuple, and now array).  You could get previously get them via `consume_list<mytype>()`, but `consume<mytype>()` didn't accept/recognize it as a list type: now it does, and similarly for dict-like types.  (This also implicitly extends to `require<T>`).

Edit: added a second commit that does a similar thing with `append()` for producers.

jagerman committed:
Allow consuming std::array's; allow other output types in consume

This rewrites the tuple/pair code a bit so that it is also usable for
std::array so that things like:

    auto arr = btdp.consume_list<std::array<int, 3>>("v");

now works.  Previously we could consume to a tuple, pair, or expandable
container like a vector, but not to an std::array.  Now we can.

It also enlarges the recognized things in `consume` to allow any
arbitrary deserializable list/dict types (including tuple, and now
array).  You could get previously get them via `consume_list<mytype>()`,
but `consume<mytype>()` didn't accept/recognize it as a list type: now
it does, and similarly for dict-like types.  (This also implicitly
extends to `require<T>`).

jagerman committed:
Allow lists/tuples for producer append

This makes `btlp.append(mylist)` and `btdp.append("key", mylist) work
when mylist is clearly a list type or tuple: both will append the value
as a sublist.

This also adds append_list() overloads for iterator pair, list, and
tuple values.

tewinget commented on oxenc/bt_producer.h:
```diff
@@ -341,6 +340,20 @@ class bt_list_producer {
     }
 
+    /// Appends an input list container as a sublist.  Equivalenet to append_list(tuple).  Note that
+    /// this is not equivalent to the iterator pair overload above: that appends to the current
```
It almost feels like we should do like python with `extend` vs `append`

jagerman replied:
Yeah, it's awkward, but I don't want to break anything that might be using the existing api.

jagerman committed:
Rename iterator-pair append to extend

- deprecated append alias for backwards compatibility
- add a list-taking version of extend()
